### PR TITLE
Enable Proguard by default on release mode

### LINKED ISSUE #11099: Enable ProGuard for Android in release mode
We should enable ProGuard for Android builds in release mode. It removes unnecessary stuff from the APK and slims down the APK size.

More information: https://developer.android.com/studio/build/shrink-code.html

Currently, enabling ProGuard on the Flutter Gallery saves about 24KB. That's not a lot, but that number could grow. For example, with ProGuard enabled, we could start depend on the Android Support Libraries in the engine without bloating our APKs as ProGuard will strip out all unused code. For the accessibility work, I'd like to depend on the Android Support Libraries because they offer much needed convenience around the raw Android APIs and also easily ensure backwards compatibility to older API versions.

/cc @jakobr-google @mravn-google @tvolkert @Hixie

### PR DESCRIPTION
**Pull request by blasten**
## Description

Add `--proguard` flag to `flutter build apk` and `flutter build appbundle` and enable it by default for release builts. 

This change makes it possible to add custom rules by placing the file `proguard-rules.pro` in the `android/` directory in case a plugin requires special rules.

### Size in bytes

| | hello_world | flutter_gallery |
|----|----|----|
|with proguard| 4,473,289 | 41,522,555 |
|without proguard| 4,524,765 |41,820,143 |

cc @Hixie 

## Related Issues

Fixes https://github.com/flutter/flutter/issues/11099

## Tests

* build_apk_test.dart

* build_appbundle_test.dart

These tests aren't hermetic yet, but these are the first ones to test the calls to Gradle.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by blasten** [comment]
Now I just realized that coverage report has false positives. The files that have decreased coverage is the result of moving code, so it can be reused.

**Comment by blasten** [comment]
Now I just realized that coverage report has false positives. The decreased coverage in some files is  the just the result of moving code, so it can be reused.

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 1]
```diff
@@ -28,11 +29,44 @@ import '../reporting/reporting.dart';
...
-final RegExp _assembleTaskPattern = RegExp(r'assemble(\S+)');
+/// Gradle utils in the current [AppContext].
+GradleUtils get gradleUtils => context.get<GradleUtils>() ?? GradleUtils.singleton();
+
+/// The implementation of Gradle Utils.
+@visibleForTesting
+class GradleUtils {
+  GradleUtils();
+
+  factory GradleUtils.singleton() {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/gradle.dart]: While we claim many things are singletons in our API docs, I don't think we've ever named a constructor `singleton` - it feels a bit leaky. Instead I would inject the GradleUtils in runInContext with the rest of the interfaces

File: packages/flutter_tools/gradle/flutter_proguard_rules.pro [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# Prevents `Fragment and FragmentActivity not found`.
+# Note: Remove once we bring the Maven dependencies.
```
Comment (dnfield) [inline packages/flutter_tools/gradle/flutter_proguard_rules.pro]: This should be a TODO, right?

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter_tools/lib/src/android/gradle.dart]: I think we also generally make a static public `instance` property for singletons, e.g. the various `Binding` classes

File: packages/flutter_tools/test/src/common.dart [hunk 3]
```diff
@@ -116,6 +116,8 @@ Future<String> createProject(Directory temp, { List<String> arguments }) async {
   final CreateCommand command = CreateCommand();
   final CommandRunner<void> runner = createTestCommandRunner(command);
   await runner.run(<String>['create', ...arguments, projectPath]);
+  // Created `.packages` since it's not created when the flag `--no-pub` is passed.
+  fs.file(fs.path.join(projectPath, '.packages')).createSync();
```
Comment (dnfield) [inline packages/flutter_tools/test/src/common.dart]: Would this not fail for users?

**Review by dnfield — COMMENTED** [review]
couple nits - was there some reason we didn't enable this in the past?  I was under the impression it could be destructive in some way for plugins that would be non-obvious for new users.  I see a way to override that for plugins here, but will that be something that (new) users will be able to understand and use correctly?

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 1 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/android/gradle.dart]: that makes sense. I almost forgot about `runInContext`. Done

File: packages/flutter_tools/gradle/flutter_proguard_rules.pro (hunk 2 above)
Comment (blasten) [reply packages/flutter_tools/gradle/flutter_proguard_rules.pro]: Done

File: packages/flutter_tools/test/src/common.dart (hunk 3 above)
Comment (blasten) [reply packages/flutter_tools/test/src/common.dart]: In the test, we pass the `--no-pub` flag to `CreateCommand`, so we don't trigger network requests. In a separate PR, it makes sense to mock the service that deals with pub, so it doesn't trigger network requests. This way, we test more code.

**Comment by blasten** [comment]
If Proguard failed, the tool will log the event and tell the user to use `--no-proguard`. 

Screen Shot 2019-09-09 at 10 37 18 AM

If we detect that a lot of folks are seeing warnings from Proguard, we could flip the default value and scoped down the issues. For instance: https://github.com/flutter/flutter/issues/37839.

**Comment by dnfield** [comment]
Wouldn't the failure potentially happen at runtime though? E.g. if proguard removed a class you actually want to use but it thought was safe to remove.

**Comment by blasten** [comment]
@dnfield I can only think of JNI calls or reflection. This won't be an issue in the engine since the rules are configured, but if a plugin is doing something similar then that plugin should use the `@Keep` annotation or provide Proguard rules. Are you aware of any plugin that might be doing something like that?  This might need better docs from our side.

**Comment by dnfield** [comment]
I'm not sure about any plugins using reflection.  I may be mis-remembering, but I had a vague notion that the way we do the plugin registration could end up having proguard shake out plugin code we want to keep.

Is there a test that covers this flag with a plugin enabled?

**Comment by blasten** [comment]
> Is there a test that covers this flag with a plugin enabled?

yes, the Flutter gallery.  I also tested manually and on Firebase device lab.

> we do the plugin registration could end up having proguard shake out plugin code we want to keep.

AFAICT, `GeneratedPluginRegistrant` imports each of the plugins and calls the `registerWith` method on the main plugin class.

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 4]
```diff
@@ -28,11 +29,37 @@ import '../reporting/reporting.dart';
 import 'android_sdk.dart';
 import 'android_studio.dart';
 
-final RegExp _assembleTaskPattern = RegExp(r'assemble(\S+)');
+/// Gradle utils in the current [AppContext].
+GradleUtils get gradleUtils => context.get<GradleUtils>();
+
+/// The implementation of Gradle Utils.
+class GradleUtils {
+  GradleUtils();
```
Comment (dnfield) [inline packages/flutter_tools/lib/src/android/gradle.dart]: uber nit: documentation for this.

**Review by dnfield — APPROVED** [review]
LGTM with nit

File: packages/flutter_tools/lib/src/android/gradle.dart (hunk 4 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/android/gradle.dart]: I made the comment in the class definition more useful.

Comment (dnfield) [reply packages/flutter_tools/lib/src/android/gradle.dart]: nice! The constructor is still missing a doc comment though.

Comment (blasten) [reply packages/flutter_tools/lib/src/android/gradle.dart]: It's an empty constructor though. Consumer of this code will use `GradleUtils` from the context most of the time.   ¯\_(ツ)_/¯ wdyt?

Comment (dnfield) [reply packages/flutter_tools/lib/src/android/gradle.dart]: You could just say that I guess.  It's public so it should have a docstring.

Comment (blasten) [reply packages/flutter_tools/lib/src/android/gradle.dart]: sg

File: packages/flutter_tools/test/general.shard/commands/build_apk_test.dart [hunk 5]
```diff
@@ -93,4 +92,205 @@ void main() {
...
+        'build-apk',
+        'proguard-failure',
+        parameters: anyNamed('parameters'),
+      )).called(1);
+    },
+    overrides: <Type, Generator>{
+      AndroidSdk: () => mockAndroidSdk,
+      GradleUtils: () => GradleUtils(),
+      FlutterProjectFactory: () => FakeFlutterProjectFactory(tempDir),
+      Logger: () => mockLogger,
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/commands/build_apk_test.dart]: testUsingContext already injects a BufferLogger called `testLogger` I think.

File: packages/flutter_tools/test/general.shard/commands/build_appbundle_test.dart [hunk 6]
```diff
@@ -76,4 +75,211 @@ void main() {
...
+        'build-appbundle',
+        'proguard-failure',
+        parameters: anyNamed('parameters'),
+      )).called(1);
+    },
+    overrides: <Type, Generator>{
+      AndroidSdk: () => mockAndroidSdk,
+      GradleUtils: () => GradleUtils(),
+      FlutterProjectFactory: () => FakeFlutterProjectFactory(tempDir),
+      Logger: () => mockLogger,
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/commands/build_appbundle_test.dart]: ditto

**Review by zanderso — APPROVED** [review]
lgtm w/ nits

File: packages/flutter_tools/test/general.shard/commands/build_apk_test.dart (hunk 5 above)
Comment (blasten) [reply packages/flutter_tools/test/general.shard/commands/build_apk_test.dart]: 👍

**Comment by rostopira** [comment]
This broke my build:

**Comment by blasten** [comment]
@rostopira what version of Gradle and the Android Gradle plugin are you using?

**Pull request merged**